router/forwarder/scanner.go

```diff
@@ -0,0 +1,106 @@
+/*
+	Copyright NetFoundry, Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package forwarder
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/openziti/fabric/ctrl_msg"
+	"github.com/openziti/fabric/pb/ctrl_pb"
+	"github.com/openziti/foundation/channel2"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+type Scanner struct {
+	ctrl        channel2.Channel
+	sessions    *sessionTable
+	interval    time.Duration
+	timeout     time.Duration
+	closeNotify <-chan struct{}
+}
+
+func NewScanner(options *Options, closeNotify <-chan struct{}) *Scanner {
+	s := &Scanner{
+		interval:    options.IdleTxInterval,
+		timeout:     options.IdleSessionTimeout,
+		closeNotify: closeNotify,
+	}
+	if s.interval > 0 {
+		go s.run()
+	} else {
+		logrus.Warnf("scanner disabled")
+	}
+	return s
+}
+
+func (self *Scanner) SetCtrl(ch channel2.Channel) {
+	self.ctrl = ch
+}
+
+func (self *Scanner) setSessionTable(sessions *sessionTable) {
+	self.sessions = sessions
+}
+
+func (self *Scanner) run() {
+	logrus.Info("started")
+	defer logrus.Warn("exited")
+
+	for {
+		select {
+		case <-time.After(self.interval):
+			self.scan()
+
+		case <-self.closeNotify:
+			return
+		}
+	}
+}
+
+func (self *Scanner) scan() {
+	sessions := self.sessions.sessions.Items()
+	logrus.Debugf("scanning [%d] sessions", len(sessions))
+
+	var idleSessionIds []string
+	for sessionId, ft := range sessions {
+		if time.Since(ft.(*forwardTable).last) > self.timeout {
+			idleSessionIds = append(idleSessionIds, sessionId)
+			logrus.Warnf("[s/%s] idle after [%s]", sessionId, self.timeout)
+		}
+	}
+
+	if len(idleSessionIds) > 0 {
+		logrus.Debugf("found [%d] idle sessions, confirming with controller", len(idleSessionIds))
+
```

Comment:
feels like self.ctrl should never be nil

Reply:
Oh, I see, scanner is initialized then ctrl is set later.

Not for this PR, but it feels like it would be cleaner to init ctrl channel first,then init router